enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go

```diff
@@ -157,41 +158,45 @@ func NewPriorityTaskScheduler(env environment.Env, exec *executor.Executor, opti
 		cpuMillisCapacity = int64(float64(resources.GetAllocatedCPUMillis()) * .80)
 	}
 
+	rootContext, rootCancel := context.WithCancel(context.Background())
 	qes := &PriorityTaskScheduler{
 		env:                   env,
 		log:                   sublog,
 		q:                     newTaskQueue(),
 		exec:                  exec,
 		newTaskSignal:         make(chan struct{}, 1),
+		rootContext:           rootContext,
 		activeTaskCancelFuncs: make(map[*context.CancelFunc]struct{}, 0),
 		shuttingDown:          false,
 		ramBytesCapacity:      ramBytesCapacity,
 		cpuMillisCapacity:     cpuMillisCapacity,
 	}
-	// This func ensures that we don't attempt to "claim" work that was enqueued
-	// but not started yet, ensuring that someone else will have a chance to get it.
+
+	// This func ensures that we don't attempt to claim work that was
+	// enqueued but not started yet, allowing another executor a chance
+	// to complete it. This is client-side "graceful" stop -- we stop
+	// processing queued work as soon as we receive a shutdown signal, but
+	// permit in-progress work to continue, up until just before the
+	// shutdown timeout, at which point we hard-cancel it.
 	env.GetHealthChecker().RegisterShutdownFunction(func(ctx context.Context) error {
 		qes.mu.Lock()
```

Comment:
It looks like the lock is being hold while waiting for the shutdown deadline.

Reply:
my bad; dashed this off before lunch without testing first oO